# Практическая работа №1

Выполнил студент гр. 0391 Каскырбай Даулет, вариант 56.

# Тема работы:

Исследование алгоритмов формирования аддитивных цепочек.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

$\quad$ Аддитивная цепочка - последовательность натуральных чисел от 1 до n
$$ 1 = a_0, a_1, ..., a_r = n ,$$ 
где каждый элемент равен сумме каких-либо двух предыдущих элементов
$$ a_{i} = a_{j} + a_{k}, k \leq j < i $$

$ l(n) $ - наименьшая длина аддитивной цепочки для заданного $ n \in \mathbb{N} $.

Неравенство для метода множителей: $ \quad l(mn) \leq l(m) + l(n) $

Неравенство для m-арного метода $ [m = 2^k, n = \sum_{j=0}^{k} d_{j}m^{t-j}] $: $ \quad l(n) = m - 2 + (k+1)t $

Неравенство для бинарного алгоритма "SX": $ \quad l(n) = \lambda(n) + \nu(n) - 1 $

### Свойства аддитивных цепочек
* Полагается, что все аддитивные цепочки - возрастающие:
$ 1 = a_0 < a_1 < ... < a_r = n $
* Если два числа из последовательности $\{a_i\}$ совпадают, то одно из них может быть опущено
* Пара $ (j, k), 0 \leq k < j < i $ называется шагом
* Если существует более одной пары $ (j, k) $, полагается, что $ j $ - наибольшее из возможных 

### Виды шагов 
* удвоение: $ \quad j = k = i - 1 $
* звездный шаг: $ \quad j = i - 1 $
* малый шаг: $ \quad \lambda(a_{i}) = \lambda(a_{i-1}) $

### Свойства шагов
* Первый шаг - всегда удвоение
* Удвоение - звездный шаг, но никогда не малый шаг
* За удвоением всегда следует звездный шаг
* Если $ i-й $ шаг не малый, то $ (i + 1)-й $ - либо звездный, либо малый либо и тот, и другой
* Если $ (i + 1)-й $ шаг ни звездный, ни малый, то $ i-й $ шаг должен быть малым

### Теорема
Если аддитивная цепочка включает $ d $ удвоений и $ f = r - d $ неудвоений, то 
$$ n \leq 2^{d-1}F_{f+3}, $$
где $ F_j $ - число Фибоначчи под номером $ j $.
### Следствие из теоремы
Если аддитивная цепочка содержит $ f $ удвоений и $ S $ малых шагов, то
$$ S \leq f \leq \frac{S}{1-log_2(\varphi)}, $$ 
где $ \varphi = \frac{\sqrt{5} + 1}{2} $ - золотое сечение.

### Алгоритм Брауэра
Для заданного числа $ n \in \mathbb{N} $ можно построить цепочку Брауэра с помощью рекуррентной формулы
$$ B_k(n) = 
\begin{cases} 
1, 2, 3, ..., 2^{k} - 1,\  если \  n < 2^{k}\\
B_k(q), 2q, ..., 2^kq, n,\  если \  n \geq 2^{k} и \  q = \lfloor{\frac{n}{2^k}} \rfloor
\end{cases} $$

Данная цепочка будет иметь длину
$$ l_B(n) = j(k + 1) + 2^k - 2 $$
при условии, что $ jk \leq \lambda(n) < (j + 1)k $

Длина цепочки будет минимизирована для больших $ n $, если 
$$ k = \lambda(\lambda(n)) - 2\lambda(\lambda(\lambda(n))) $$

#### Уточнения
* Задается фикированный параметр $ k $. Выполняются вычисления вспомогательных чисел:
$$ 
d = 2^k, q_1 = [\frac{n}{d}], r_1 = n \ mod \ d \Rightarrow n = q_{1}d + r_{1} \ (0 \leq r_{1} < d) \\
q_2 = [\frac{q_1}{d}], r_2 = q_1 \ mod \ d \Rightarrow q_1 = q_{2}d + r_{2}
$$
* Данная процедура выполняется до тех пор, пока $ q_s $ не станет меньше $ d $. Следовательно $ q_{s-1} = q_{s}d + r_{s} $
* Таким образом,
$$ n = 2^kq_1 + r_1 = ... = 2^k(2^k(...(2^kq_s + r_s)...) + r_2) + r_1 $$
$$ B_k(n): 1, 2, 3, ..., 2^k - 1, 2q_s, 4q_s, ..., 2^kq_s + r_s, ..., n $$

### Алгоритм Яо
* Имеет ту же вычислительную сложность, что и алгоритм Брауэра
* Раскладываем число $ n \in \mathbb{N} $ в $ 2^k-ой $ системе счисления:
$$ n = \sum_{i=0}^j a_{j}2^{ik}, \ a_j \neq 0 $$
* Введем функцию $ d(z) $:
$$ d(z) = \sum_{i: a_i = z} 2^{ik} $$

#### Ход алгоритма
* Задаем базовую последовательность: $ 1, 2, 4, ..., 2^{\lambda(n)} $
* Вычисляем значения $ d(z) $ для $ z \in \{ 1, 2, 3, ..., 2^k - 1 \} $, причем $ d(z) \neq 0 $
* Вычисляем $ zd(z) $ для всех $ z $ 
* В конечном итоге получаем
$$ n = \sum_{z = 1}^{2^k - 1} zd(z) $$

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.


## Порядок выполнения работы
1.  Вручную (т.е. не реализовывая алгоритм на Sage) построить последовательность вычислений бинарным методом и методом множителей для $𝑥^𝑛$ для 2-3 значений 𝑛 (значение 𝑛 > 30 выбираются студентом самостоятельно). Сравнить количество операций для каждого метода, сделать выводы.

2. Реализовать алгоритм Яо для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра 𝑘, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.

3. Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений 𝑛 > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.

4. Проверить гипотезу Шольца–Брауэра для всех натуральных 1 $\leqslant$ 𝑛 $\leqslant$ 12 на алгоритме дробления вектора индексов. Результаты оформить в виде таблицы. Сделать выводы.

# Пункт 1
**Построим последовательность вычислений бинарным методом для $x^{31}$**

| i 	| N 	| Y 	| Z 	|
|---	|---	|---	|---	|
| 0 	| 31 	| 1 	| x 	|
| 1 	| 15 	| x 	| x^2 	|
| 2 	| 7 	| x^3 	| x^4 	|
| 3 	| 3 	| x^7 	| x^8 	|
| 4 	| 1 	| x^15 	| x^16 	|
| 5 	| 0 	| x^31 	| x^16 	|

Вычислим число операций для возведения в степень $x^{31}$:

Двоичная запись числа 31: 11111

$l(31)$ = $\lambda(31)$ + $\nu(31)$ - 1 = 4 + 5 - 1 = 8

 **Построим последовательность вычислений бинарным методом для $x^{32}$**

| i 	| N 	| Y 	| Z 	|
|---	|---	|---	|---	|
| 0 	| 32 	| 1 	| x 	|
| 1 	| 16 	| 1 	| x^2 	|
| 2 	| 8 	| 1 	| x^4 	|
| 3 	| 4 	| 1 	| x^8 	|
| 4 	| 2 	| 1 	| x^16 	|
| 5 	| 1 	| 1 	| x^32 	|
| 6 	| 0 	| x^32 	| x^32 	|

Вычислим число операций для возведения в степень $x^{31}$:

Двоичная запись числа 31: 11111

$l(32)$ = $\lambda(32)$ + $\nu(32)$ - 1 = 5 + 1 - 1 = 5

Теперь построим последовательность вычислений методом множителей для $x^{31}$ и $x^{32}$


$x^{31}$ = $x^{30}*x$. 

$n$ = 30 = $p*q$ , $p$ - наименьший простой делитель $\Rightarrow$ $p$ = 2

$x^{30}$ = $(x^{15})^2$ = $(x*x^{14})^2$ = $(x*(x^7)^2)^2$ = $(x*(x*(x^3)^2)^2)^2$ = $(x*(x*(x*(x^2))^2)^2)^2$ - 7 операций $\Rightarrow$ возведение в 31 степень произойдет за 8 операций.


$x^{32}$ = $(x^{16})^2$ = $((x^{8})^2)^2$ = $(((x^{4})^2)^2)^2$ = $(((((x^2)^2)^2)^2)^2$ - 5 операций

**Сравним результаты**

| n 	| Бинарный метод 	| Метод множителей 	|
|---	|---	|---	|
| 31 	| 8 	| 8 	|
| 32 	| 5 	| 5 	|

**Вывод:**

Бинарный метод работает медленнее. Но так как мы использовали небольшие значения $n$, то количество операций в бинарном методе и методе множителей одинаково. 

# Реализация алгоритма Яо. Пункт 2.

Теперь реализуем алгоритм Яо для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра k.

In [20]:
def YAO(n, k):
    syst = []
    additive_chain = []
    temp_n = n
    p = 2**k
    z = []
    syst.append([N, 0])
    while temp_n >= p:
        syst.append([temp_n//p, temp_n % p])
        z.append(temp_n % p)
        temp_n //= p
    z.append(syst[len(syst) - 1][0])
    power = z
    z = list(set(z))
    z.sort()
    print("z =", z)
    d = []
    d_z = []
    for i in z:
        d_i = 0;
        for j in range(0, len(power)):
            if i == power[j]:
                d_i += (2**k)**j
        d.append(d_i)
        d_z.append(d_i*i)
    print("d(z) = ", d)
    print("z*d(z) = ", d_z)
    equal = 0
    for i in d_z:
        equal += i
    print(equal)
    additive_chain = []
    first = int(log(n)/log(2))
    additive_chain.append(1)
    for i in range(1, first + 1):
        additive_chain.append(additive_chain[i - 1]*2)
    print(additive_chain)

In [21]:
n = int(input())
k = int(input())
YAO(n, k)

31415
3
z = [2, 5, 6, 7]
d(z) =  [64, 512, 8, 4097]
z*d(z) =  [128, 2560, 48, 28679]
31415
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384]


31415

3

z = [2, 5, 6, 7]

d(z) =  [64, 512, 8, 4097]

z*d(z) =  [128, 2560, 48, 28679]

31415

[1, 2, 4, 8, 16, 24, 32, 48, 64, 128, 256, 512, 1024, 1536, 2048, 2560, 4096, 4097, 8192, 8194, 12291, 16384, 20485, 28679, 28807, 31367, 31415]

# Алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа.

In [7]:
def hamming(n):
    n = Integer(n)
    return n.popcount()
def generate_vec(m, ch_part):
    vec1 = []
    vec2 = []
    for i in range(1, m + 1):
        if i <= int(ch_part):# do not forgot to fix it
            vec1.append(i)
        else:
            vec2.append(i)
    return [vec1, vec2]
def get_num(vec):
    line = []
    line.append(1)
    for i in range(len(vec)):
        line.append(line[i] + line[vec[i] - 1])
    return line[len(line) - 1]
# decremenation
def dec(vec):
    _len = len(vec)
    for i in range(_len):
        if vec[_len - i - 1] > 1:
            vec[_len - i - 1] -= 1
            for j in range(_len - i, _len):
                vec[j] = j + 1
            break
def generate_part(_len, first):
    part = [0]*_len
    for i in range(_len):
        part[i] = first + i
    return part
def get_chain(vec):
    line = []
    line.append(1)
    for i in range(len(vec)):
        line.append(line[i] + line[vec[i] - 1])
    return line
def empty(vec):
    for i in vec:
        if i > 1:
            return False
    return True
def index(n):
    l_min = math.ceil(math.log2(n)) #minimum length
    l_max = int(math.log2(n) + hamming(n)) #maximum length
    for m in range(l_min, l_max):
        q = int(m/2)
        vec = generate_vec(m, q)
        #print(vec)
        static_part = generate_part(q, 1)
        
        while true: #sorting out all static parts
            dynamic_part = generate_part(m - q, q + 1)
            bounds = [0]*2
            a = get_num(static_part)
            bounds[0] = a + m - q
            bounds[1] = a * 2 ** (m - q)

            
            if n < bounds[0] or n > bounds[1]:
                if empty(static_part) and len(static_part) > 1:
                    break
                dec(static_part)
                continue
                
            while True: #sorting out all dynamic parts
                if get_num(static_part + dynamic_part) == n:
                    return static_part + dynamic_part
                if empty(dynamic_part):
                    break
                dec(dynamic_part)
                #print(static_part + dynamic_part)
            if empty(static_part):
                break
            dec(static_part)   
    print("Problem")
import timeit
def calculate(N):
    start_time = timeit.default_timer()
    res = index(N)
    print("completed in", timeit.default_timer() - start_time, "seconds")
    return res
n = int(input()) #get a number to create our vector
ind = calculate(n)
print("Vector", ind)
print("Chain", get_chain(ind))

1203
completed in 159.36330470000394 seconds
Vector [1, 1, 3, 4, 5, 6, 7, 8, 9, 10, 10, 3, 7]
Chain [1, 2, 3, 6, 12, 24, 48, 96, 192, 384, 768, 1152, 1155, 1203]


1000

completed in 1.724324999959208 seconds

Vector [1, 2, 3, 4, 4, 1, 7, 8, 9, 10, 11, 10]

Chain [1, 2, 4, 8, 16, 24, 25, 50, 100, 200, 400, 800, 1000]


1025

completed in 0.00034009991213679314 seconds

Vector [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1]

Chain [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1025]


1023

completed in 168.57069039996713 seconds

Vector [1, 1, 3, 4, 5, 4, 7, 8, 9, 10, 11, 8, 3]

Chain [1, 2, 3, 6, 12, 24, 30, 60, 120, 240, 480, 960, 1020, 1023]


1488

completed in 262.1193876999896 seconds

Vector [1, 2, 3, 4, 5, 6, 5, 8, 9, 10, 11, 9, 6, 5]

Chain [1, 2, 4, 8, 16, 32, 64, 80, 160, 320, 640, 1280, 1440, 1472, 1488]


1203

completed in 159.36330470000394 seconds

Vector [1, 1, 3, 4, 5, 6, 7, 8, 9, 10, 10, 3, 7]

Chain [1, 2, 3, 6, 12, 24, 48, 96, 192, 384, 768, 1152, 1155, 1203]


# Проверим теорему Шольца-Брауэра

In [ ]:
for i in range(1, 13):
    k = index(i)
    _k = index(2**i - 1)
    chain = get_chain(k)
    _chain = get_chain(_k)
    print("n =", i, ", l(n) =", len(chain), ", l(2^n - 1) =", len(_chain))
    print("l(2^n - 1) <= n - 1 + l(n): ", len(_chain) <= i - 1 + len(chain))
    print()

n = 1 , l(n) = 1 , l(2^n - 1) = 1
l(2^n - 1) <= n - 1 + l(n):  True

n = 2 , l(n) = 2 , l(2^n - 1) = 3
l(2^n - 1) <= n - 1 + l(n):  True

n = 3 , l(n) = 3 , l(2^n - 1) = 5
l(2^n - 1) <= n - 1 + l(n):  True

n = 4 , l(n) = 3 , l(2^n - 1) = 6
l(2^n - 1) <= n - 1 + l(n):  True

n = 5 , l(n) = 4 , l(2^n - 1) = 8
l(2^n - 1) <= n - 1 + l(n):  True

n = 6 , l(n) = 4 , l(2^n - 1) = 9
l(2^n - 1) <= n - 1 + l(n):  True

n = 7 , l(n) = 5 , l(2^n - 1) = 11
l(2^n - 1) <= n - 1 + l(n):  True

n = 8 , l(n) = 4 , l(2^n - 1) = 11
l(2^n - 1) <= n - 1 + l(n):  True

n = 9 , l(n) = 5 , l(2^n - 1) = 13
l(2^n - 1) <= n - 1 + l(n):  True

n = 10 , l(n) = 5 , l(2^n - 1) = 14
l(2^n - 1) <= n - 1 + l(n):  True



| n 	| l*(n) 	| l*(2^n-1) 	| Theorem 	|
|---	|---	|---	|---	|
| 1 	| 1 	| 1 	| completed 	|
| 2 	| 2 	| 3 	| completed 	|
| 3 	| 3 	| 5 	| completed 	|
| 4 	| 3 	| 6 	| completed 	|
| 5 	| 4 	| 8 	| completed 	|
| 6 	| 4 	| 9 	| completed 	|
| 7 	| 5 	| 11 	| completed 	|
| 8 	| 4 	| 11 	| completed 	|
| 9 	| 5 	| 13 	| completed 	|
| 10 	| 5 	| 14 	| completed 	|
| 11 	| 6 	| 16 	| completed 	|
| 12 	| 5 	| 16 	| completed 	|

# Выводы:

$\bullet$ Время работы алгоритма очень большое. Зависит от длины числа. Полученыые цепочки не являются кратчайшими. 

$\bullet$ Гипотеза Шольца-Брауэра верна для n $\in$ (1;12)



